Contact force sensors

camevor opened the pull request:
## Description
Contact force sensors

## Newton Migration Guide

Please ensure the migration guide for **warp.sim** users is up-to-date with the changes made in this MR.

- [ ] The migration guide in ``docs/migration.rst`` is up-to date

## Before your PR is "Ready for review"

- [x] All commits are [signed-off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt--s) to indicate that your contribution adheres to the [Developer Certificate of Origin](https://developercertificate.org/) requirements
- [x] Necessary tests have been added and new examples are tested (see `newton/tests/test_examples.py`)
- [ ] Documentation is up-to-date
- [x] Code passes formatting and linting checks with `pre-commit run -a`


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced a GPU-accelerated contact sensor system for querying and aggregating contact forces between entities in physics simulations.
  * Added an example simulation demonstrating contact sensing with articulated models and randomized controls.
  * Enabled mapping and conversion of contact data between MuJoCo and Newton physics representations.

* **Bug Fixes**
  * None.

* **Tests**
  * Added unit tests for contact sensor functionality, verifying force aggregation and contact detection.
  * Integrated a new example-based test for the contact sensor simulation.

* **Documentation**
  * None.

* **Chores**
  * None.

* **Refactor**
  * None.

* **Style**
  * None.

* **Revert**
  * None.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

shi-eric commented:
Hey @camevor, you can make your Newton org membership public so you don't need manual approval on the GPU job:

1. Go to the organization's People page: https://github.com/orgs/newton-physics/people
2. Find your username in the list.
3. Click the dropdown next to your name and change your visibility from 'Private' to 'Public'.

camevor commented:
Thank you, @shi-eric !

eric-heiden commented on newton/sim/builder.py:
```diff
@@ -3262,6 +3279,69 @@ def add_face(i, j, k):
+            contact_partners_shape: pattern to match contact partner shape names; one entity per matching shape.
+            contact_partners_body: pattern to match contact partner body names; one entity per matching body.
+            match_fun: function taking a name and a pattern and returning true if the name matches the pattern;
+            defaults to fnmatch. Accepts "re" for `re.match()`.
```
I would remove the `"re"` option - the user could just pass some lambda function that uses `re.match()` themselves. The Selection API defaults to `fnmatch` also.

eric-heiden commented on newton/utils/contact_sensor.py:
```diff
@@ -0,0 +1,397 @@
+):
+    """Populate ContactInfo object from the solver or from the Contacts object."""
+    if solver is not None:
+        if isinstance(solver, MuJoCoSolver):
```
I think there should be no dependency on a particular solver in this module.
We could add an abstract method to `SolverBase.update_contact_info(model, constacts, contact_info)` that does nothing by default and `MuJoCoSolver` implements that with what you have already. "update_newton_contacts" is not a good name I think because you are updating `ContactInfo`, not `newton.Contacts`. So consider naming it `update_contact_info`.

eric-heiden commented on newton/utils/contact_sensor.py:
```diff
@@ -0,0 +1,397 @@
+        """Entities for the sensors in the query, n_sensors"""
+        self.contact_partner_entities: list[Entity] = None
+        """Entities for the contact partners in the query, n_contact_partners"""
+        self.entity_pairs: np.ndarray = None  # entity pair matrix
```
We have a typedef for Numpy arrays in `newton.core.types` called `nparray` because the correct type alias for Numpy arrays is more complicated than this.

eric-heiden commented on newton/utils/contact_sensor.py:
```diff
@@ -0,0 +1,397 @@
+        self.entity_pairs: np.ndarray = None  # entity pair matrix
+        """Pairs of sensor and contact partner indices for the query, shape (n_sensors, n_contact_partners, 2)"""
+
+    def finalize(
```
Generally `finalize()` in Newton means you allocate things on the device which is passed to this method, see `Mesh.finalize()` or `ModelBuilder.finalize()`. Here you are copying things to the `ContactView` instance, so it looks like another constructor? Or consider renaming the method to `update()` or something more descriptive.

eric-heiden commented on newton/utils/contact_sensor.py:
```diff
@@ -0,0 +1,397 @@
+    def eval_contact_sensors(self, contact_info: ContactInfo):
+        self.contact_reporter._select_aggregate_net_force(contact_info)
+
+    def finalize(self):
```
Same as before, maybe rename this to `evaluate()` or something more descriptive.

eric-heiden commented on newton/utils/contact_sensor.py:
```diff
@@ -0,0 +1,397 @@
+        self.n_entity_pairs: int = len(entity_pairs)
+        self._create_sp_ep_arrays(entity_pairs)
+        # net force (1 vec3 per entity pair)
+        self.net_force = wp.zeros(self.n_entity_pairs, dtype=wp.vec3)
```
This allocation and the ones in `_create_sp_ep_arrays` need a device specified.

eric-heiden commented on newton/sim/builder.py:
```diff
@@ -3591,8 +3671,85 @@ def finalize(self, device: Devicelike | None = None, requires_grad: bool = False
+
             return m
 
+    def _create_sensor(
```
Do we need this separate method? It also seems very specific to `add_contact_sensor`, can't we move this code there?

eric-heiden commented on newton/solvers/mujoco/solver_mujoco.py:
```diff
@@ -1469,6 +1524,65 @@ def color_collision_shapes(model: Model, selected_shapes: nparray, visualize_gra
+        if contact_info.force is None:
+            contact_info.force = wp.empty(nconmax, dtype=wp.float32, device=model.device)
+
+        num_threads = min(mj_data.nconmax, 8192)
```
Does 8192 have to be the same number as `newton.utils.contact_sensor.NUM_THREADS`? Then you should probably use that.

nvlukasz reviewed (commented):
I like the functionality in this PR, being able to tract contacts between both bodies and shapes. I think the code is quite clean, but I agree with other comments that it could be simplified further when it comes to the API. More functions, less classes is the Newtonic way :)

mmacklin commented on newton/sim/builder.py:
```diff
@@ -3622,5 +3779,79 @@ def find_shape_contact_pairs(self, model: Model):
         model.shape_contact_pairs = wp.array(np.array(contact_pairs), dtype=wp.vec2i, device=model.device)
         model.shape_contact_pair_count = len(contact_pairs)
+
+    def _get_entities(
```
Agreed, `ModelBuilder` shouldn't have any access to the `Model`.

mmacklin commented on newton/examples/example_selection_articulations.py:
```diff
@@ -245,6 +261,11 @@ def step(self):
                 self.simulate()
+
+        convert_contact_info(self.model, self.contact_info, self.solver)
+        self.model.eval_contact_sensors(self.contact_info)
```
I would really like to find a design where we don't have to introduce the `ContactInfo` class at all, it has too much overlap with the existing `Contacts` class and feels like it should be more of an implementation detail.

mmacklin commented:
Thanks Chris, in addition to the other fine-grained comments, a few higher-level things to discuss:

I wonder contact sensors really belongs on the `Model` at all - similar to the selection APIs `ArticulationView`, I wonder if if these `ContactView` objects could be created in a standalone way, e.g.: could we have a workflow similar to this:

```
model = builder.finalize()
solver = newton.solvers.MuJoCoSolver(model)

sensor1 = ContactSensor(model, "foot1", "ground")
sensor2 = ContactSensor(model, "foot2", "ground")

for i in range(N):

   # run collision detection
   contacts = model.collide(state)

   # each solver responsible for updating the total force at each contact
   solver.step(state, contacts)

   # update sensors
   eval_contact_sensor(sensor1, contacts)
   eval_contact_sensor(sensor2, contacts)

   # query total force
   sensor1.get_total_force()
   sensor2.get_total_force()
```

The goal would be to introduce as few new types as possible, and decouple them as much as possible.

camevor commented:
Thank you for your comments, Miles!

I agree that decoupling the contact sensors from the model is a cleaner design.
The main reason for this choice was the goal of evaluating all contact sensors in a single pass over the contact array, as the model is a convenient vehicle for transparently tracking several contact sensors. 
In this design, the `ContactView` is just a slice of the contact force matrix that contains the results of a `add_contact_sensor()` declaration, since the contact force matrix is shared by all sensors.

Given the complexity arising from this, we might want to make the assumption that the number of different contact sensors we expect users to define is too small to prioritize making a single pass over the contacts? This would simplify things quite a bit, at the cost of somewhat worse scaling behavior in the number of contact sensor definitions.

camevor commented:
I've reworked this PR to incorporate all your comments @eric-heiden @nvlukasz @mmacklin .

Summary of points addressed in this rework:

* Stand-alone `ContactSensor` class, fully de-coupled from `ModelBuilder`
* Contact sensor definitions are no longer batched; sensors are initialized immediately
* Abandoned the "Entity" concept, use bodies, shapes and `body_shapes` instead
* Moved solver-dependent contact update logic to solver class, added abstract `SolverBase.update_contact_info`
* Removed `ContactView` class, use `ContactSensor` class instead
* Removed `ContactInfo` class, filling `Contacts` from MuJoCo solver as an interim solution
* Removed `ContactReporter` class, fused with `ContactViewManager` into `ContactSensor`
* Removed `ContactViewManager`
* Removed `ContactQuery` dataclass since it's no longer needed
* Removed body/shape confusion hint, leaving it to users to implement this check
* Removed `EntityNotFoundError`
* Removed `SentinelMeta`
* Removed type alias for sets of shapes
* Several minor (better naming, defining devices, types, etc.)

Below are points that are not addressed in this version and other problematic items:
* Designing a better interface for solvers to provide internal state, e.g. forces or contacts.
* Consistent handling of the `Contacts` class. Since it is still likely to undergo major changes, I added a single commit that removes the `ContactInfo` class and uses the `Contacts` class instead with its own fields, as Lukasz suggested.
* Sharing `binary_search` with broadphase.
* The matching logic for Contact Sensors could serve as a template for the Selection API. If this approach is well-received, we can easily unify the two interfaces in a future PR.
* Bespoke example for the Contact Sensors: I have something fitting in mind here, but to unblock this PR, there is a placeholder example simply to show how the API can be used.

I'll create follow-up issues for the points in the 'not addressed' section after this is merged.
